### DISCUSSION TITLE
Mixed Content Error

### DISCUSSION THREAD

**Discussion by LinuxSquare** [zitadel/zitadel#5060] [Q&A] [answered]
Hi there Zitadel-Community (and maybe Devs)

I've discovered Zitadel on the `awesome to self-host` list and though, I'd throw a look into it.
When hosted as `127.0.0.1 / ::1` service, it works as intended.

Now I went a step further and followed the guide, to deploy it as a production service with docker:
https://zitadel.com/docs/self-hosting/manage/configure#passing-the-configuration

I've copied the docker-compose and the three other `.yaml` config-files and changed them up a little, so they would work with my domain.
I've also added the additional config, regarding the `external domain`:
```yaml
ExternalSecure: true
ExternalDomain: 'zitadel.mydomain.ch'
ExternalPort: 443
```

The docker-instance is also behind a reverse proxy running with [`nginx proxy manager`](https://nginxproxymanager.com/)
My reverse proxy is also hidden behind a reverse proxy of my provider.

The connection from the first reverse proxy to my reverse proxy is secured via a self-signed ssl-certificate.
Connections coming from the internet are secured with a Let's Encrypt SSL certficiate, that's hosted on the reverse proxy of my provider.

Now when I startup the service on my machine (Guest OS: Debian 11, Host: Proxmox) it works flawlessly.
But when I fire up a web-browser with the domain, it loads like a quintillion of years.

Logs in the console say:
```
Mixed Content: The page at 'https://zitadel.mydomain.ch/ui/console/' was loaded over HTTPS, but requested an insecure XMLHttpRequest endpoint 'http://zitadel.mydomain.ch/.well-known/openid-configuration'. This request has been blocked; the content must be served over HTTPS.
```

In my reverse proxy, I've enabled `Force SSL` , `HTTP/2`and `HSTS` aswell as `HSTS Subdomains`
![image](https://user-images.githubusercontent.com/39629672/212769217-cda5aaa0-5e56-4e16-9578-5c0aec085ad5.png)

Here are my configuration files:
- [docker-compose.yaml](https://pastebin.com/i1DLR0DW)
- [zitadel-config.yaml](https://pastebin.com/K1TnT6fe)
- [zitadel-init-steps.yaml](https://pastebin.com/0NF6u4NV)
- [zitadel-secrets.yaml](https://pastebin.com/iSsEw4X5)

I know, the values are still default. Those will be changed when going in production.

Thank you in advance for any help I can get.

Greetings from your neighbouring canton.
LinuxSquare

**Comment by fforootd** [comment, marked as answer]
Hm, that's a good question 😁 

Can you by any means provide me a [HAR file from your browser](https://support.google.com/admanager/answer/10358597?hl=en) when access the mentioned url?

I want to check how the browser invokes the requests.

> Greetings from your neighbouring canton.

Greetings back, hope you got some snow as well 😄

**Reply by LinuxSquare** [reply]
Sure, shall I send it to you like via pn or something similar?  
Because inside the `.har` file you'll find my complete domain and I don't want that everyone can download the file and see what's the real domain. (To be as secure as possible and prevent like ddos / dos attacks :P)

> Greetings back, hope you got some snow as well :smile: 

Jup, we've got some snow aswell :rofl:

**Reply by fforootd** [reply]
> Sure, shall I send it to you like via pn or something similar?

Just drop me a mail on florian@zitadel.com that works best

**Reply by LinuxSquare** [reply]
> Just drop me a mail on [florian@zitadel.com](mailto:florian@zitadel.com) that works best

Okay, is sent. Thank you very much :smile:

**Reply by fforootd** [reply]
Great, will look into this today!

**Reply by fforootd** [reply]
I found something. Console is trying to call .well-known/openid-configuration endpoint in your har file without htttps.

Since this endpoint is important things cannot proceed 😁 

Usually this error has something to do if ZITADEL is not properly aware that TLS is set to external.

**Reply by fforootd** [reply]
Found it 😁 

```
  command: 'start-from-init --config /example-zitadel-config.yaml --config /example-zitadel-secrets.yaml --steps /example-zitadel-init-steps.yaml --masterkey "${ZITADEL_MASTERKEY}" --tlsMode disabled'
```

`--tlsMode disabled` is the issue switch this to `--tlsMode external`

**Reply by LinuxSquare** [reply]
Oh my god. I think I'mma go eat a broom now. :broom: 
I was so sure, that I tested it with `--tlsMode external` :cry: 

Thank you very much for taking your time to fix my stupid mistake.
That's customer care at it's finest :+1: 

But one more question to go.
The login happened, but now this came up in the Network Tab:
![image](https://user-images.githubusercontent.com/39629672/213542944-4b8b75e9-920d-4a59-a995-e657945e1ad2.png)
It doesn't look very healthy to me.

They're pretty 404 Errors
![image](https://user-images.githubusercontent.com/39629672/213543258-22a7e06f-96d3-445e-95e0-9346ee23fc88.png)


Do you have any idea, what it could be? :smiling_face_with_tear: 

Thank you

**Reply by fforootd** [reply]
that sounds like http2 is not working properly

**Reply by LinuxSquare** [reply]
Hi @fforootd 
I took again a look into a fresh .har file, and the requests are being processed using HTTP/1.1
```json
"request": {
   "method": "GET",
   "url": "https://zitadel.mydomain.ch/ui/console/users/me",
   "httpVersion": "HTTP/1.1",
```
...
```json
"response": {
   "status": 200,
   "statusText": "OK",
   "httpVersion": "HTTP/1.1",
```

It almost looks like, that Nginx Proxy Manager doesn't support HTTP/2 quite well, even though I've checked it in the options, as seen the original request.
I have seen something in the issues of Zitadel (#5054), mentioning HTTP/1.1 support, since some providers do not support HTTP/2 as being mentioned in the issue.

Is that still a thing or have the plans been borked? 
If not, is there like roadmap of a possible release? (Maybe 2025 - 2026?)

Thank you in advance.

**Reply by fforootd** [reply]
Yeah the h1 issue is something we are currently investigating.

is there a way in NPM to force the upstream connection to use h2?

Because with plain nginx that can be done this way:

```
http {
    server {
        listen 443;

        ssl_certificate     ssl/certificate.pem;
        ssl_certificate_key ssl/key.pem;

        location / {
            grpc_pass grpc://localhost:8080;
            grpc_set_header Host $host;
        }
    }
}
```

**Reply by LinuxSquare** [reply]
Hi there @fforootd 

I've managed to force http/2 into the container, thanks to your config.
There's actually a hidden text-box in npm, where you can configure the grpc_pass options.

![image](https://user-images.githubusercontent.com/39629672/214130075-63d8c4b0-94e3-4cfa-9bec-3b2676f441c9.png)

Thank you very much for your help!

Now, I won't disturb you anymore and start testing with my Zitadel Setup.
A thousand thanks from Moschtindie :stuck_out_tongue_closed_eyes:

**Reply by fforootd** [reply]
Ah wonderful, nice to see that this works with NPM.

Always happy to lend a hand! 

Greetings from Bratwurstcity 😁

**Reply by LinuxSquare** [reply]
I've made a little summary for everyone else in the future who may stumble across the same problem. :smile: 
See [here](https://github.com/zitadel/zitadel/discussions/5060#discussioncomment-4760283)

**Reply by fforootd** [reply]
Thank you!

**Comment by LinuxSquare** [comment]
Now a quick summary for those who have a similar setup like mine:
Your setup looks like:
* You're running [Nginx Proxy Manager](https://nginxproxymanager.com/) as a reverse proxy and it works as intended.
* (Optional) In front of your NPM instance, is another reverse proxy.
* Zitadel is running inside a docker-container, setup like described in [the official guide](https://zitadel.com/docs/self-hosting/manage/configure).

How to solve this issue?

Step `I have another reverse proxy in front of mine` 1)
Make sure the connection from the first reverse proxy to your reverse proxy is ssl secured.
No need to issue a real certificate from any CA or a Let's Encrypt cert for **YOUR** NPM-Instance. A self-signed certificate is enough, because the first reverse proxy is the one, who's exposed to the internet. Only this one needs a ca signed certificate (Let's Encrypt, Verisign, Digicert Quovadis etc.), so the connection from the net is marked as "trusted".

OR

Step `My reverse proxy is directly exposed to the internet` 1)
If your reverse proxy is directly exposed to the internet, just use your wildcard certificate or ca-cert you've ordered from your CA or just let NPM generate a Let's Encrypt certificate for you (see image).
![image](https://user-images.githubusercontent.com/39629672/214135392-ef447e3f-4758-457e-8d2a-4e36eb95d63d.png)


Step 2)
Check inside the docker-compose.yml if `--tlsMode` is set to `external`
[See @fforootd's answer](https://github.com/zitadel/zitadel/discussions/5060#discussioncomment-4725759)

Step 3)
Preferably, enable `Force SSL` and `HTTP/2 Support` in the `SSL`-tab, when you're editing your Proxy Host inside NPM.
![Screenshot_20230123_203340](https://user-images.githubusercontent.com/39629672/214134494-58d9cbb9-4598-4371-8bc0-5e7209945ac4.png)


Step 4)
Under the tab `Custom locations`, add a new location, by clicking on `Add Location`.
Then define in `location` the value `/`
and in `scheme` select `http`, set the IPv4 of the host and the port, where the Zitadel-instance is hosted.
Now, click on the cogwheel-icon and add following config, as mentioned [here](https://github.com/zitadel/zitadel/discussions/5060#discussioncomment-4754695):
```
grpc_pass grpc://YOUR-IP:PORT
grpc_set_header Host $host;
```

Otherwise, set your values, according to the upcoming image:
![image](https://user-images.githubusercontent.com/39629672/214132551-f4883094-114a-4013-b2b5-5e843998163a.png)

Here in this image, exchange Port `8080` with your configured port.
